twenty-first/src/util_types/merkle_tree.rs

```diff
@@ -179,15 +201,18 @@ impl MerkleTree {
         Ok(nodes)
     }
 
-    /// Given a list of leaf indices, return the indices of exactly those nodes that
-    /// are needed to prove (or verify) that the indicated leafs are in the Merkle
-    /// tree.
-    // This function is not defined as a method (taking self as argument) since it's
-    // needed by the verifier, who does not have access to the Merkle tree.
-    fn authentication_structure_node_indices(
-        num_leafs: usize,
-        leaf_indices: &[usize],
-    ) -> Result<impl ExactSizeIterator<Item = usize> + use<>> {
+    /// Compute the node indices for an authentication structure.
+    ///
+    /// Given a list of leaf indices, return the indices of exactly those nodes
+    /// that are needed to prove (or verify) that the indicated leafs are in the
+    /// Merkle tree.
+    ///
+    /// Returns an error if any of the leaf indices is bigger than the number of
+    /// leafs.
+    pub fn authentication_structure_node_indices(
+        num_leafs: MerkleTreeLeafIndex,
+        leaf_indices: &[MerkleTreeLeafIndex],
+    ) -> Result<impl ExactSizeIterator<Item = MerkleTreeLeafIndex> + use<>> {
```

Comment:
This is the reason I'm not a fan of creating aliases for simple types: the function name and the documentation imply `MerkleTreeNodeIndex` for the iterator's item type, yet it is `MerkleTreeLeafIndex`.

Reply:
Good catch! This should absolutely be `MerkleTreeNodeIndex`.

I understand the drawback of not having the compiler's help to identify inconsistencies. It leads to wrong aliases slipping through the cracks, like here. But if that's a compelling argument against using type aliases, then it's also a compelling argument against comments in the code describing what's going on. Do you have a similar policy on comments (not including doctests or docstrings)?

Reply:
To a certain degree. If the function's signature (including name) fully explains what the function does, then I don't add the noise that a comment would now be. However, since the comment has to follow fewer rules than the name of the function, it's easy to add information not captured in the signature, which makes adding a comment worthwhile often enough.

All that said, I might misunderstand what you're getting at, since you are excluding docstrings from consideration. The type aliases are public, as should the description of a function's “what” be, meaning it should go in the docstring.

Reply:
Type aliases can have docstrings, and these particular docstrings can be used to document index conventions. This solution strikes me as better than the alternative, which is to document the relevant bits of the index conventions in the functions' docstrings. This alternative leads to a scattered convention documentation.